File: python/src/heaton_life/core/decimal_text.py
"""The decimal-string grammar of viewport centers (spec/deep-zoom.md "Viewport contract").

One grammar for every consumer: Viewport validation, the fixed-point orbit parse
(core.bignum), and the float64 projection T0 renders from. The C# port scans the
same way (HeatonLife.DecimalText), so a string one side accepts the other does too.

    [ws] [+|-] digits-with-at-most-one-point [ (e|E) [+|-] digits ] [ws]

At least one digit before the exponent (".5" and "5." are fine, "." is not); at most
10,000 digits before it; exponent magnitude at most 100,000; ASCII digits only;
surrounding ASCII spaces, tabs, CRs and LFs are ignored. No NaN, no infinities, no
"_" separators. The two bounds keep parsing and the orbit precision finite for any
string a caller can hand in.
"""

from __future__ import annotations

import math

MAX_DIGITS = 10_000
MAX_EXPONENT = 100_000
_WHITESPACE = " \t\r\n"


def scan(text: str) -> tuple[bool, int, int]:
    """(negative, digits, net_exponent) with value = digits * 10^net_exponent.

    Raises ValueError for anything outside the grammar.
    """
    s = text.strip(_WHITESPACE)
    if not s:
        raise ValueError("empty decimal string")
    i = 0
    negative = False
    if s[i] in "+-":
        negative = s[i] == "-"
        i += 1
    digits = 0
    digit_count = 0
    fraction_digits = 0
    saw_digit = saw_point = saw_exponent = False
    while i < len(s):
        c = s[i]
        if "0" <= c <= "9":
            digit_count += 1
            if digit_count > MAX_DIGITS:
                raise ValueError(f"more than {MAX_DIGITS} digits: {text[:40]!r}...")
            digits = digits * 10 + (ord(c) - 48)
            if saw_point:
                fraction_digits += 1
            saw_digit = True
        elif c == "." and not saw_point:
            saw_point = True
        elif c in "eE" and saw_digit:
            i += 1
            saw_exponent = True
            break
        else:
            raise ValueError(f"not a decimal number: {text!r}")
        i += 1
    if not saw_digit:
        raise ValueError(f"not a decimal number: {text!r}")

    exponent = 0
    if saw_exponent and i >= len(s):
        raise ValueError(f"truncated exponent: {text!r}")
    if i < len(s):
        exp_negative = False
        if s[i] in "+-":
            exp_negative = s[i] == "-"
            i += 1
        if i >= len(s):
            raise ValueError(f"truncated exponent: {text!r}")
        while i < len(s):
            c = s[i]
            if not "0" <= c <= "9":
                raise ValueError(f"not a decimal number: {text!r}")
            exponent = exponent * 10 + (ord(c) - 48)
            if exponent > MAX_EXPONENT:
                raise ValueError(f"exponent out of range: {text!r}")
            i += 1
        if exp_negative:
            exponent = -exponent
    return negative, digits, exponent - fraction_digits


def to_float(text: str) -> float:
    """The float64 nearest the decimal value: one rounding, ties to even.

    Subnormals round like everything else, magnitudes past the float64 range become
    infinities (IEEE-754 round-to-nearest), and "-0" is -0.0. CPython's float()
    rounds a decimal string correctly (David Gay's algorithm), so after the grammar
    check it IS the exact conversion; the C# port computes it from the digits.
    """
    scan(text)
    return float(text.strip(_WHITESPACE))


def difference(minuend: str, subtrahend: str) -> float:
    """The float64 nearest ``minuend - subtrahend``, computed exactly from the digits and
    rounded once (ties to even; infinities past the float64 range; an exact zero is
    +0.0). The off-center reference's offset (spec/deep-zoom.md "Off-center
    reference") -- never a difference of two already-rounded doubles.
    """
    neg_a, digits_a, exp_a = scan(minuend)
    neg_b, digits_b, exp_b = scan(subtrahend)
    common = min(exp_a, exp_b)
    a = digits_a * 10 ** (exp_a - common) * (-1 if neg_a else 1)
    b = digits_b * 10 ** (exp_b - common) * (-1 if neg_b else 1)
    diff: int = a - b
    if diff == 0:
        return 0.0
    try:
        if common >= 0:
            return float(diff * 10**common)
        denominator: int = 10**-common
        return diff / denominator
    except OverflowError:
        return float("inf") if diff > 0 else float("-inf")


def difference_x(minuend: str, subtrahend: str) -> tuple[float, int]:
    """``minuend - subtrahend`` as floatexp (m, e): the exact difference of the digits
    rounded once to a 53-bit mantissa (ties to even), with no exponent range limit --
    difference() for T2, where the off-center offset lies far below float64
    (spec/deep-zoom.md "T2"). An exact zero is (0.0, 0)."""
    neg_a, digits_a, exp_a = scan(minuend)
    neg_b, digits_b, exp_b = scan(subtrahend)
    common = min(exp_a, exp_b)
    a = digits_a * 10 ** (exp_a - common) * (-1 if neg_a else 1)
    b = digits_b * 10 ** (exp_b - common) * (-1 if neg_b else 1)
    diff: int = a - b
    if diff == 0:
        return 0.0, 0
    numerator = abs(diff) * (10**common if common >= 0 else 1)
    denominator = 10**-common if common < 0 else 1
    # A 53-bit quotient: q = floor(value * 2^s) in [2^52, 2^53).
    s = 52 - (numerator.bit_length() - denominator.bit_length())
    while True:
        num, den = (numerator << s, denominator) if s >= 0 else (numerator, denominator << -s)
        q, r = divmod(num, den)
        if q >= 1 << 53:
            s -= 1
        elif q < 1 << 52:
            s += 1
        else:
            break
    if 2 * r > den or (2 * r == den and q & 1):
        q += 1
        if q == 1 << 53:
            q >>= 1
            s -= 1
    mantissa = math.ldexp(float(q), -52)  # exact: q has 53 bits
    return (-mantissa if diff < 0 else mantissa), 52 - s


def digits_of(value: int) -> str:
    """``str(value)`` for a non-negative int of any length. CPython refuses ``str()`` of
    an int past 4,300 digits by default, and the grammar allows 10,000: longer values
    are split at a power of ten and each half converted."""
    if value.bit_length() <= 13_000:  # under 3,914 digits
        return str(value)
    half = value.bit_length() * 30_103 // 200_000  # about half the digit count
    high, low = divmod(value, 10**half)
    return digits_of(high) + digits_of(low).rjust(half, "0")


def format_scaled(value: int, places: int) -> str:
    """``value * 10^-places`` written positionally with exactly ``places`` fraction digits
    (none, and no point, when ``places`` is 0): "-" only for a nonzero negative, no "+",
    one "0" before the point when the magnitude is below 1. Raises ValueError when the
    result would carry more than MAX_DIGITS digits -- it could not be read back.
    """
    if places < 0:
        raise ValueError(f"places must be non-negative, got {places}")
    # 10^MAX_DIGITS has 33,220 bits: refuse a longer value before converting it.
    if places >= MAX_DIGITS or abs(value).bit_length() > 33_220:
        raise ValueError(f"more than {MAX_DIGITS} digits in a positional decimal")
    magnitude = digits_of(abs(value)).rjust(places + 1, "0")
    if len(magnitude) > MAX_DIGITS:
        raise ValueError(f"more than {MAX_DIGITS} digits in a positional decimal")
    body = f"{magnitude[:-places]}.{magnitude[-places:]}" if places else magnitude
    return f"-{body}" if value < 0 else body


def positional(text: str) -> str:
    """The same value written positionally: no exponent, no "+", no sign on zero, no
    leading zeros beyond one -- and the same number of decimal places the precision
    rule counts (max(-net_exponent, 0), spec/deep-zoom.md), so an orbit's working bits
    do not change. "1e-5" -> "0.00001", "-1.2E-7" -> "-0.00000012", "2.5E1" -> "25",
    "0.10" -> "0.10". Raises ValueError outside the grammar, or when the positional
    form would exceed MAX_DIGITS digits ("1e-20000" has 20,000 places).
    """
    negative, digits, net_exponent = scan(text)
    signed = -digits if negative else digits
    if net_exponent >= 0:
        if len(digits_of(digits)) + net_exponent > MAX_DIGITS:
            raise ValueError(f"more than {MAX_DIGITS} digits in a positional decimal")
        return format_scaled(signed * 10**net_exponent, 0)
    return format_scaled(signed, -net_exponent)

File: dotnet/src/HeatonLife.Core/DecimalText.cs
using System;
using System.Globalization;
using System.Numerics;

namespace HeatonLife
{
    /// <summary>
    /// The decimal-string grammar of viewport centers (spec/deep-zoom.md "Viewport
    /// contract"), and exact conversions out of it. One grammar for every consumer —
    /// <see cref="Viewport"/> validation, the fixed-point orbit parse
    /// (<see cref="ReferenceOrbit"/>), and the float64 projection T0 renders from — and
    /// the same one the Python reference scans (core/decimal_text.py):
    ///
    ///     [ws] [+|-] digits-with-at-most-one-point [ (e|E) [+|-] digits ] [ws]
    ///
    /// At least one digit before the exponent (".5" and "5." are fine, "." is not); at
    /// most 10,000 digits before it; exponent magnitude at most 100,000; ASCII digits
    /// only; surrounding ASCII spaces, tabs, CRs and LFs are ignored. No NaN, no
    /// infinities. The two bounds keep parsing and the orbit precision finite for any
    /// string a caller can hand in. Hand-rolled because Core takes no Regex.
    /// </summary>
    internal static class DecimalText
    {
        internal const int MaxDigits = 10000;
        internal const int MaxExponent = 100000;

        private static readonly char[] Whitespace = { ' ', '\t', '\r', '\n' };

        /// <summary>
        /// Split a decimal string into sign, digits and net exponent (value = digits *
        /// 10^netExponent); ArgumentException for anything outside the grammar.
        /// </summary>
        internal static void Scan(string text, out bool negative, out BigInteger digits, out int netExponent)
            => ScanCore(text, true, out negative, out digits, out netExponent);

        /// <summary>
        /// The net exponent alone (value = digits * 10^netExponent), validating the same
        /// grammar but building no BigInteger: what the precision rule and the Viewport's
        /// validation need, on every frame, allocation-free.
        /// </summary>
        internal static int NetExponent(string text)
        {
            ScanCore(text, false, out _, out _, out int netExponent);
            return netExponent;
        }

        private static void ScanCore(
            string text, bool buildDigits, out bool negative, out BigInteger digits, out int netExponent)
        {
            if (text == null)
                throw new ArgumentNullException(nameof(text));
            string s = text.Trim(Whitespace);
            if (s.Length == 0)
                throw new ArgumentException("empty decimal string", nameof(text));
            int i = 0;
            negative = false;
            if (s[i] == '+' || s[i] == '-')
            {
                negative = s[i] == '-';
                i++;
            }

            digits = BigInteger.Zero;
            int digitCount = 0;
            int fractionDigits = 0;
            bool sawDigit = false;
            bool sawPoint = false;
            bool sawExponent = false;
            for (; i < s.Length; i++)
            {
                char c = s[i];
                if (c >= '0' && c <= '9')
                {
                    if (++digitCount > MaxDigits)
                        throw new ArgumentException($"more than {MaxDigits} digits", nameof(text));
                    if (buildDigits)
                        digits = digits * 10 + (c - '0');
                    if (sawPoint)
                        fractionDigits++;
                    sawDigit = true;
                }
                else if (c == '.' && !sawPoint)
                {
                    sawPoint = true;
                }
                else if ((c == 'e' || c == 'E') && sawDigit)
                {
                    i++;
                    sawExponent = true;
                    break;
                }
                else
                {
                    throw new ArgumentException($"not a decimal number: '{text}'", nameof(text));
                }
            }
            if (!sawDigit)
                throw new ArgumentException($"not a decimal number: '{text}'", nameof(text));

            int exponent = 0;
            if (sawExponent && i >= s.Length)
                throw new ArgumentException($"truncated exponent: '{text}'", nameof(text));
            if (i < s.Length)
            {
                bool expNegative = false;
                if (s[i] == '+' || s[i] == '-')
                {
                    expNegative = s[i] == '-';
                    i++;
                }
                if (i >= s.Length)
                    throw new ArgumentException($"truncated exponent: '{text}'", nameof(text));
                for (; i < s.Length; i++)
                {
                    char c = s[i];
                    if (c < '0' || c > '9')
                        throw new ArgumentException($"not a decimal number: '{text}'", nameof(text));
                    exponent = exponent * 10 + (c - '0');
                    if (exponent > MaxExponent)
                        throw new ArgumentException($"exponent out of range: '{text}'", nameof(text));
                }
                if (expNegative)
                    exponent = -exponent;
            }
            netExponent = exponent - fractionDigits;
        }

        /// <summary>
        /// The double nearest a decimal string's exact value: one rounding, ties to
        /// even, subnormals like everything else, infinities past the float64 range,
        /// and "-0" as -0.0 — what CPython's correctly rounded float() gives. Computed
        /// from the digits because <c>double.Parse</c> is only guaranteed correctly
        /// rounded on .NET Core 3.0+, and Unity runs Core on Mono and IL2CPP.
        /// </summary>
        internal static double ToDouble(string text)
        {
            Scan(text, out bool negative, out BigInteger digits, out int netExponent);
            if (digits.IsZero)
                return negative ? -0.0 : 0.0;
            return netExponent >= 0
                ? RatioToDouble(negative, digits * BigInteger.Pow(10, netExponent), BigInteger.One)
                : RatioToDouble(negative, digits, BigInteger.Pow(10, -netExponent));
        }

        /// <summary>
        /// The double nearest <paramref name="minuend"/> − <paramref name="subtrahend"/>,
        /// computed exactly from the digits and rounded once (ties to even; infinities past
        /// the float64 range; an exact zero is +0.0) — the off-center reference's offset
        /// (spec/deep-zoom.md "Off-center reference"), never a difference of two
        /// already-rounded doubles. The Python reference's decimal_text.difference.
        /// </summary>
        internal static double Difference(string minuend, string subtrahend)
        {
            Scan(minuend, out bool negA, out BigInteger a, out int expA);
            Scan(subtrahend, out bool negB, out BigInteger b, out int expB);
            int common = Math.Min(expA, expB);
            if (negA)
                a = -a;
            if (negB)
                b = -b;
            BigInteger diff = a * BigInteger.Pow(10, expA - common) - b * BigInteger.Pow(10, expB - common);
            if (diff.IsZero)
                return 0.0;
            bool negative = diff.Sign < 0;
            BigInteger magnitude = negative ? -diff : diff;
            return common >= 0
                ? RatioToDouble(negative, magnitude * BigInteger.Pow(10, common), BigInteger.One)
                : RatioToDouble(negative, magnitude, BigInteger.Pow(10, -common));
        }

        /// <summary>
        /// <paramref name="value"/> × 10^-<paramref name="places"/> written positionally
        /// with exactly <paramref name="places"/> fraction digits (none, and no point, when
        /// it is 0): "-" only for a nonzero negative, no "+", one "0" before the point when
        /// the magnitude is below 1 (spec/navigation.md "Conventions"). Throws when the
        /// result would carry more than <see cref="MaxDigits"/> digits — it could not be
        /// read back. The Python reference's decimal_text.format_scaled.
        /// </summary>
        internal static string FormatScaled(BigInteger value, int places)
        {
            if (places < 0)
                throw new ArgumentOutOfRangeException(nameof(places), "places must be non-negative");
            // 10^MaxDigits has 33,220 bits: refuse a longer value before formatting it.
            if (places >= MaxDigits || BitLength(BigInteger.Abs(value)) > 33220)
                throw new ArgumentException($"more than {MaxDigits} digits in a positional decimal");
            // Format the magnitude: a culture's negative sign never enters the digits.
            string magnitude = BigInteger.Abs(value).ToString(CultureInfo.InvariantCulture).PadLeft(places + 1, '0');
            if (magnitude.Length > MaxDigits)
                throw new ArgumentException($"more than {MaxDigits} digits in a positional decimal");
            string body = places > 0
                ? magnitude.Substring(0, magnitude.Length - places) + "." + magnitude.Substring(magnitude.Length - places)
                : magnitude;
            return value.Sign < 0 ? "-" + body : body;
        }

        /// <summary>
        /// The same value written positionally, with the same number of decimal places the
        /// precision rule counts (max(-net exponent, 0)) — no exponent, no "+", no sign on
        /// zero. The Python reference's decimal_text.positional.
        /// </summary>
        internal static string Positional(string text)
        {
            Scan(text, out bool negative, out BigInteger digits, out int netExponent);
            BigInteger signed = negative ? -digits : digits;
            if (netExponent >= 0)
            {
                if (digits.ToString(CultureInfo.InvariantCulture).Length + netExponent > MaxDigits)
                    throw new ArgumentException($"more than {MaxDigits} digits in a positional decimal");
                return FormatScaled(signed * BigInteger.Pow(10, netExponent), 0);
            }
            return FormatScaled(signed, -netExponent);
        }

        /// <summary>
        /// ±numerator/denominator (both positive) to the nearest double, ties to even,
        /// with IEEE-754's subnormal and overflow behavior.
        /// </summary>
        internal static double RatioToDouble(bool negative, BigInteger numerator, BigInteger denominator)
        {
            // The value lies in (2^(e-1), 2^(e+1)); its floor(log2) is e or e - 1.
            int e = BitLength(numerator) - BitLength(denominator);
            bool atLeast = e >= 0
                ? numerator >= denominator << e
                : numerator << -e >= denominator;
            int floorLog2 = atLeast ? e : e - 1;
            if (floorLog2 >= 1024)
                return negative ? double.NegativeInfinity : double.PositiveInfinity;
            if (floorLog2 < -1075)
                return negative ? -0.0 : 0.0;                  // below half the smallest subnormal

            // Round once, at the result's ulp: 53 significant bits while normal,
            // 2^-1074 below.
            int ulp = Math.Max(floorLog2 - 52, -1074);
            BigInteger n = ulp >= 0 ? numerator : numerator << -ulp;
            BigInteger d = ulp >= 0 ? denominator << ulp : denominator;
            BigInteger q = BigInteger.DivRem(n, d, out BigInteger remainder);
            int versusHalf = (remainder << 1).CompareTo(d);
            if (versusHalf > 0 || (versusHalf == 0 && !q.IsEven))
                q += BigInteger.One;
            return Compose(negative, q, ulp);
        }

        /// <summary>
        /// ±q * 2^ulp assembled from its IEEE-754 bit pattern, where q &lt;= 2^53 is
        /// already rounded at that ulp (normal when q &gt;= 2^52, else subnormal with
        /// ulp = -1074). No floating-point operation runs, so nothing rounds twice.
        /// </summary>
        internal static double Compose(bool negative, BigInteger q, int ulp) => Compose(negative, (long)q, ulp);

        /// <summary><see cref="Compose(bool, BigInteger, int)"/> for a mantissa already in a long.</summary>
        internal static double Compose(bool negative, long q, int ulp)
        {
            long mantissa = q;
            if (mantissa == 1L << 53)
            {
                mantissa >>= 1;                                // rounding carried into the next binade
                ulp++;
            }
            long pattern;
            if (mantissa >= 1L << 52)
            {
                int biased = ulp + 52 + 1023;
                if (biased >= 2047)
                    return negative ? double.NegativeInfinity : double.PositiveInfinity;
                pattern = ((long)biased << 52) | (mantissa & ((1L << 52) - 1));
            }
            else
            {
                pattern = mantissa;                            // subnormal (or zero): exponent field 0
            }
            double result = BitConverter.Int64BitsToDouble(pattern);
            return negative ? -result : result;
        }

        /// <summary>Position of the highest set bit of a positive BigInteger.</summary>
        internal static int BitLength(BigInteger value)
        {
            byte[] bytes = value.ToByteArray();               // little-endian, two's complement
            int i = bytes.Length - 1;
            while (i > 0 && bytes[i] == 0)
                i--;
            int top = bytes[i];
            int inTop = 0;
            while (top > 0)
            {
                inTop++;
                top >>= 1;
            }
            return i * 8 + inTop;
        }
    }
}
